### ISSUE TITLE
getForm-Hook wird ausschließlich im Insert-Tag {{getForm::$id}} getriggert.

### ISSUE THREAD

**Issue by Tastaturberuf** [contao/contao#2678]
https://github.com/contao/contao/blob/4.7.6/core-bundle/src/Resources/contao/library/Contao/Controller.php#L541
https://github.com/contao/contao/blob/4.7.6/core-bundle/src/Resources/contao/library/Contao/InsertTags.php#L512

Ich wollte ein Bundle bauen in dem man nach erfolgreichen Absenden eines Formulars, statt diesem Content-Elemente anzeigen kann. Um nicht immer eine ähnliche Weiterleitungsseite erstellen zu müssen. Leider wird der Hook bei regulärer Nutzung vom Form-Content-Element nicht ausgelöst.

**Comment by aschempp** [comment]
Hört sich ein bisschen nach https://extensions.contao.org/?p=terminal42%2Fcontao-ajaxform an, vielleicht kannst du da abschauen/kopieren?

**Comment by fritzmg** [comment]
Sort of related to https://github.com/contao/contao/issues/1764

When a form is inserted as a content element, only the `getContentElement` hook will be fired. If it is inserted as a module via the page layout, only the `getFrontendModule` hook will be fired. So if you want to cover all bases you need to implement the `getContentElement`, `getFrontendModule` and `getForm` hook.

**Comment by Tastaturberuf** [comment]
Ok, verstehe ich. Dann sollte das auch noch am Hook erklärt werden. Jetzt erscheint mir der Hook aber recht sinnlos.

**Comment by fritzmg** [comment]
Der Hook an sich ist ok. Er wird immer dann aufgerufen, wenn `Controller::getForm` aufgerufen wird, genau wie `getContentElement` oder `getFrontendModule`.

**Comment by Tastaturberuf** [comment]
Ja aber das passiert eben **nur** wenn ein Formular im als Insert-Tag eingebunden wird. Und das ist halt nicht der reguläre Fall.

**Comment by Tastaturberuf** [comment]
Ich versuche jetzt in einem eigenen Listener mit `Controller::getForm` in `getFrontendModule` und `getContentElement` zu implementieren wenn es sich um Formulare handelt, damit der Hook quasi immer gefeuert wird.

**Comment by fritzmg** [comment]
> Ja aber das passiert eben **nur** wenn ein Formular im als Insert-Tag eingebunden wird. Und das ist halt nicht der reguläre Fall.

Ja, schon klar. Aber das hat nichts direkt mit dem Hook zu tun, sondern mit dem generellen Konzept wie Elemente wie Formulare, Inhaltselemente und Module von Contao erzeugt werden.

**Comment by Tastaturberuf** [comment]
Ja hat es nicht, aber der Hook verspricht eben mehr als er tut. Ich schick nen PR für die Docs.

**Comment by fritzmg** [comment]
Can you provide that clarification for all hooks? ;) See also https://github.com/contao/docs/issues/433

**Comment by Tastaturberuf** [comment]
Jetzt wird’s wieder kompliziert für mal eben ne Änderung machen ;)

**Comment by Tastaturberuf** [comment]
Man kann das Verhalten für Formulare relativ einfach fixen:

```php
class FixGetFormInsertTagListener
{
    /**
     * @Hook("getContentElement")
     */
    public function getContentElement(ContentModel $contentModel, string $buffer, object $element): string
    {
        if ( $element instanceof Form )
        {
            return Controller::getForm($element->getModel());
        }

        if ( 'module' === $element->type )
        {
            $moduleModel = ModuleModel::findByPk($element->module);

            if ( $moduleModel->type === 'form' )
            {
                $formModel = FormModel::findByPk($moduleModel->form);

                return Controller::getForm($formModel);
            }
        }

        return $buffer;
    }


    /**
     * @Hook("getFrontendModule")
     */
    public function getFrontendModule(ModuleModel $moduleModel, string $buffer, object $module): string
    {
        if ( $module instanceof Form )
        {
            return Controller::getForm($module->getModel());
        }

        return $buffer;
    }
}
```

Könnte man für die anderen Fälle eventuell auch machen. Is halt eher nen Workround.

**Comment by fritzmg** [comment]
Well it's not really a workaround, it's what you have to do currently. As I said, for content elements only the `getContentElement` hook will be executed, since only `Controller::getContentElement` will be excuted for a content element.

**Comment by fritzmg** [comment]
I do agree though that the situation is very inconsistent. For example, the content elements

* `alias`
* `module`
* `form`

do not use

* `Controller::getContentElement()`
* `Controller::getFrontendModule()`
* `Controller::getForm()`

respectively. Whereas the content element

* `article`

on the other hand **does** use

* `Controller::getArticle`

So while it is impossible to capture every front end module or form with the `getFrontendModule` or `getForm` hook alone, it is possible to capture the output of every article via the `getArticle` hook.

So to fix/change this behavior, `ContentAlias` and `ContentModule` would need to be changed so that they use `Controller::getContentElement` and `Controller::getFrontendModule` respectively, instead of instantiating the element or module on their own - and we would need to introduce a new `ContentForm` content element and `ModuleForm` module (which would use `Controller::getForm`), instead of using the `Form` class as a content element and front end module directly.

However, that would probably be a BC break.

**Comment by leofeyer** [comment]
As discussed in Mumble on February 11th, we cannot really change the current behavior without breaking BC, so all we can do is to make it more clear in the documentation. @fritzmg has already done this in contao/docs#433.

**Comment by aschempp** [comment]
I guess this can be closed then?

**Comment by fritzmg** [comment]
Any `incompatible` tickets need to stay open for the next major version, do they not?

**Comment by leofeyer** [comment]
> So to fix/change this behavior, `ContentAlias` and `ContentModule` would need to be changed so that they use `Controller::getContentElement` and `Controller::getFrontendModule` respectively

As discussed in the Contao call, we want to change this as described.

> and we would need to introduce a new `ContentForm` content element and `ModuleForm` module (which would use `Controller::getForm`), instead of using the `Form` class as a content element and front end module directly.

I will look into this as well.
